[PS-1850] Product Search API

TxyShannon opened the pull request.

TxyShannon committed:
updating swift sdk with product search functions

TxyShannon committed:
testing response

TxyShannon committed:
Added the newer response params, objects, groupresults, sysmeta

TxyShannon committed:
comments and testing latest parameters

TxyShannon committed:
Update ViProductSearchTest.swift

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+
+    public var vaDidmd5 : String? = nil
+
+    public var vaN1 : Int? = nil
```
wrong data type, this is numeric which include double

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+
+    public var vaN1 : Int? = nil
+
+    public var vaN2 : Int? = nil
```
wrong data type

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+            dict["debug"] = debug ? "true" : "false"
+        }
+        
+        if !customParams.isEmpty {
```
move this to the bottom of the method after all other parameters

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+            dict["va_app_version"] = val
+        }
+        
+        if vaAaid != nil {
```
inconsistent, should be if let vaAaid = vaAaid

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+            dict["va_aaid"] = vaAaid
+        }
+        
+        if vaDidmd5 != nil {
```
change to if let

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+            dict["va_didmd5"] = vaDidmd5
+        }
+        
+        if vaN1 != nil {
```
change to if let

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+        }
+        
+        if vaN1 != nil {
+            dict["va_n1"] = vaN1
```
= String(vaN1)

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+        }
+        
+        if vaN2 != nil {
+            dict["va_n2"] = vaN2
```
String(vaN2)

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViBaseProductSearchParam.swift:
```diff
@@ -0,0 +1,285 @@
+            dict["va_s1"] = vaS1
+        }
+        
+        if vaS2 != nil {
```
if let

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViSeachByIdParam.swift:
```diff
@@ -0,0 +1,38 @@
+    
+    /// Product ID is actually used as part of the query path and not a normal parameter variable
+    public var productId : String? = nil
+    
```
missing `return_product_info` param

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViSearchByImageParam.swift:
```diff
@@ -0,0 +1,116 @@
+    /// - returns: Compressed jpegData
+    public func getImageData() -> Data? {
+        if let image = image {
+            return image.jpegData(compressionQuality: 0.97)
```
compression quality need to be configurable (high/medium/low), take a reference of existing code for visearch

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViSearchByImageParam.swift:
```diff
@@ -0,0 +1,116 @@
+    public override func toDict() -> [String: Any] {
+        var dict = super.toDict()
+        
+        if imUrl != nil {
```
change all this to if let style

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViSearchByImageParam.swift:
```diff
@@ -0,0 +1,116 @@
+        }
+        
+        if image != nil {
+            dict["image"] = getImageData()
```
need to make sure to test search with binary image properly in test case

thehung111 replied:
do u really need to pass the binary object into the dictionary? by default it will probably appear in the query string. Have you test?

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViSearchByImageParam.swift:
```diff
@@ -0,0 +1,116 @@
+            dict["box"] = "\(b.x1),\(b.y1),\(b.x2),\(b.y2)"
+        }
+        
+        if detection != nil {
```
if let

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViSearchByImageParam.swift:
```diff
@@ -0,0 +1,116 @@
+        }
+        
+        if detectionLimit > 0 {
+            dict["detection_limit"] = detectionLimit
```
String(detectionLimit)

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ProductSearch/ViSearchByImageParam.swift:
```diff
@@ -0,0 +1,116 @@
+            dict["detection_sensitivity"] = detectionSensitivity
+        }
+        
+        dict["search_all_objects"] = searchAllObjects
```
searchAllObjects ? "true" : "false"

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Request/ViRequestSerialization.swift:
```diff
@@ -21,6 +21,13 @@ open class ViRequestSerialization {
     
+    
+    /// generate the url with query string and escape parameter properly
+    public func generateRequestUrl( baseUrl: String , apiEndPoint: String , searchParams:[String: Any]) -> String {
```
should have another enum here e.g. PSSearchEndpoints or ViProductSearchEndpoints

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Response/ProductSearch/ViProductSearchResponse.swift:
```diff
@@ -0,0 +1,206 @@
+                }
+                
+                if let box = dict["box"] as? [Int] {
+                    item.box = ViBox(x1: box[0], y1: box[1], x2: box[2], y2: box[3])
```
need to check that box has at least 4 items

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/ViProductSearch.swift:
```diff
@@ -0,0 +1,192 @@
+    /// - parameter appKey: App key
+    /// - parameter placementId: Placement ID
+    public func setUp(appKey:String, placementId:Int) -> Void {
+        self.appKey = appKey
```
this code need to call the other setup method, instead of repeating the code

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/ViProductSearch.swift:
```diff
@@ -0,0 +1,192 @@
+    ///
+    /// - parameter appKey: App key
+    /// - parameter placementId: Placement ID
+    public func setUp(appKey:String, placementId:Int) -> Void {
```
is this a static function? the constructor is private, how are the users calling this?

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Response/ProductSearch/ViProductSearchResponse.swift:
```diff
@@ -0,0 +1,206 @@
+            }
+            
+            if let objs = json["objects"] as? [Any] {
+                objects = ViResponseData.parseObjectResults(objs)
```
cannot reuse this method, the result format between visearch and product search are different

thehung111 replied:
this object results need to call parseProductResults internally

thehung111 commented on ViSearchSDK/ViSearchSDK/Classes/Response/ProductSearch/ViProductSearchResponse.swift:
```diff
@@ -0,0 +1,206 @@
+                let item = ViGroupResult(group: group)!
+                
+                if let imageResults = dict["results"] as? [Any] {
+                    item.results = ViResponseData.parseResults(imageResults)
```
cannot reuse, API response is different. ViSearch is having value_map, Product Search is `data`

thehung111 reviewed (changes requested):
pls test properly , use `dump` to print object content.
The error is so critical. You did not parse the result correctly for `group_by` and `search_all_objects`

TxyShannon committed:
Updating according to PR comments, specifically checked for group_by, search_all_object, facets, in the test class

TxyShannon committed:
Updated test cases and readme

thehung111 committed:
update doc

thehung111 committed:
bump version

thehung111 commented on ViSearchSDK/ViSearchSDKTests/ViProductTest.swift:
```diff
@@ -0,0 +1,109 @@
+    func parseMethod(text: String) -> ViProduct? {
+        do {
+            let data = text.data(using: .utf8)!
+            let dict = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as! Dictionary<String, Any>
```
this is not the right way to test, you suppose to call the actual parse method in the response class and assert. By doing this, you are not verifying whether your implementation is correct. In fact this is a new implementation.

thehung111 committed:
add config example

thehung111 committed:
add point param

thehung111 committed:
fix resize box bug

thehung111 committed:
rename methods

thehung111 committed:
update comment

thehung111 committed:
update version

thehung111 committed:
add test

thehung111 reviewed (approved):
fixed box resize issue after image is resize. updated docs.

everything looks good, we will merge and publish after the API is formally released.

TxyShannon committed:
Updated test cases

Removed parsing of the json data in the test class and use the static parser in the actual response class instead.
